spark/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,298 @@
+import org.apache.spark.sql.sources.v2.reader.streaming.Offset;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A mirco-batch based Spark Structured Streaming reader for Iceberg table. It will track the added
+ * files and generate tasks per batch to process newly added files. By default it will process
+ * all the newly added files to the current snapshot in each batch, user could also set this
+ * configuration "max-files-per-trigger" to control the number of files processed per batch.
+ */
+class StreamingReader extends Reader implements MicroBatchReader {
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingReader.class);
+  private static final int DEFAULT_MAX_FILES_PER_TRIGGER = 1000;
+
+  private static final Comparator<FileScanTask> COMPARATOR = (left, right) -> {
+    int pathCompare = left.file().path().toString().compareTo(right.file().path().toString());
+    int startCompare = Long.compare(left.start(), right.start());
+    int lengthCompare = Long.compare(left.length(), right.length());
+    return pathCompare == 0 ? (startCompare == 0 ? lengthCompare : startCompare) : pathCompare;
+  };
+
+  private StreamingOffset startOffset;
+  private StreamingOffset endOffset;
+
+  private final Table table;
+  private final int maxFilesPerTrigger;
+  private final long splitSize;
+  private final int lookback;
+  private final long openFileCost;
+  private final Long startSnapshotId;
+
+  /**
+   * Utility class to track the snapshotId, task and index of task within this snapshot.
+   */
+  @VisibleForTesting
+  static class IndexedTask {
+    private final long snapshotId;
+    private final FileScanTask task;
+    private final int index;
+
+    IndexedTask(long snapshotId, FileScanTask task, int index) {
+      this.snapshotId = snapshotId;
+      this.task = task;
+      this.index = index;
+    }
+
+    long snapshotId() {
+      return snapshotId;
+    }
+
+    FileScanTask task() {
+      return task;
+    }
+
+    int index() {
+      return index;
+    }
+  }
+
+  StreamingReader(Table table, Broadcast<FileIO> io, Broadcast<EncryptionManager> encryptionManager,
+                  boolean caseSensitive, DataSourceOptions options) {
+    super(table, io, encryptionManager, caseSensitive, options);
+
+    this.table = table;
+    this.maxFilesPerTrigger =
+      options.get("max-files-per-trigger").map(Integer::parseInt).orElse(DEFAULT_MAX_FILES_PER_TRIGGER);
+    Preconditions.checkArgument(maxFilesPerTrigger > 0,
+        "Option max-files-per-trigger '%d' should > 0", maxFilesPerTrigger);
+
+    this.startSnapshotId = options.get("starting-snapshot-id").map(Long::parseLong).orElse(null);
+    if (startSnapshotId != null) {
+      if (!SnapshotUtil.ancestorOf(table, table.currentSnapshot().snapshotId(), startSnapshotId)) {
+        throw new IllegalStateException("The option starting-snapshot-id " + startSnapshotId +
+          "is not a valid snapshot id");
```

Comment:
I think this should be more specific, like "is not an ancestor of the current snapshot". If this says it isn't a valid snapshot, but it appears in the `snapshots` list then it would be confusing.